File: Python/727. Minimum Window Subsequence.py
class Solution:
    def minWindow(self, S: str, T: str) -> str:
        m, n = len(T), len(S)
        dp = [[0 for i in range(n + 1)] for j in range(m + 1)]
        for j in range(n + 1):
            dp[0][j] = j + 1
        for i in range(1, m + 1):
            for j in range(1, n + 1):
                if T[i - 1] == S[j - 1]:
                    dp[i][j] = dp[i - 1][j - 1]
                else:
                    dp[i][j] = dp[i][j - 1]            
        start = 0
        max_len = n + 1        
        for j in range(1, n + 1):
            if dp[m][j] != 0:
                if j - dp[m][j] + 1 < max_len:
                    start = dp[m][j]
                    max_len = j - dp[m][j] + 1
        return "" if max_len == n + 1 else S[start - 1:start + max_len - 1]

File: 2021/727. Minimum Window Subsequence.cpp
class Solution {
public:
    string minWindow(string S, string T) {
        int m = S.length(), n = T.length();
        vector<vector<int>> dp(m, vector<int>(n, m));
        int minStart = m, minLen = m;
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (S[i] == T[j]) {
                    if (i != 0 && j != 0) {
                        dp[i][j] = min(dp[i - 1][j], dp[i - 1][j - 1]) + 1;
                    } else {
                        dp[i][j] = (j == 0) ? 1 : m;
                    }                   
                    if (j == n - 1 && dp[i][j] < minLen) {                        
                        minLen = dp[i][j];
                        minStart = i - dp[i][j] + 1;
                    }
                } else {
                    dp[i][j] = (i == 0) ? m : dp[i - 1][j] + 1;
                }
            }
        }
        return S.substr(minStart, minLen);
    }
};

class Solution {
public:
    string minWindow(string S, string T) {
        string res;
        int j = 0;
        for (int i = 0; i < S.length(); i++) {
            if (S[i] == T[j]) {
                j++;
                if (j >= T.length()) {
                    int end = i + 1;
                    j--;
                    while (j >= 0) {
                        if (S[i] == T[j]) {
                            j--;
                        }
                        i--;
                    }
                    i++;
                    string tmp = S.substr(i, end - i);
                    if (res.length() == 0 || tmp.length() < res.length()) {
                        res = tmp;
                    }
                    j = 0;
                }
            }
        }
        return res;
    }
};